File: strings/0038-count-and-say.py
import unittest

# https://leetcode.com/problems/count-and-say/

# python3 -m unittest strings/0038-count-and-say.py


class Solution(unittest.TestCase):
    # # Approach: Recursive
    # # Time: O(nxm), m=max(len(RLE))
    # # Space: O(n)
    # def countAndSay(self, n: int) -> str:
    #     if n == 1: return "1"
    #     s = self.countAndSay(n-1)
    #     rle, count = "", 1
    #     for idx in range(1, len(s)):
    #         if s[idx] != s[idx-1]:
    #             rle = f"{rle}{count}{s[idx-1]}"
    #             count = 0
    #         count += 1
    #     return f"{rle}{count}{s[-1]}"

    # Approach: Iterative
    # Time: O(n)
    # Space: O(n)
    def countAndSay(self, n: int) -> str:
        rle = "1"
        for _ in range(n - 1):
            new_rle, count = "", 1
            for idx in range(1, len(rle)):
                if rle[idx] != rle[idx - 1]:
                    new_rle = f"{new_rle}{count}{rle[idx-1]}"
                    count = 0
                count += 1
            rle = f"{new_rle}{count}{rle[-1]}"
        return rle

    def test(self):
        for n, expected in [
            (4, "1211"),
            (1, "1"),
        ]:
            output = self.countAndSay(n)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: strings/0038-count-and-say.go
package strings

import (
	"strconv"
	"strings"
)

// https://leetcode.com/problems/count-and-say/

// RLE: Run-length encoding

// Approach: Iterative
// Time: O(nxm), m=max(len(RLE))
// Space: O(n)
func countAndSay(n int) string {
	rle := "1"

	var newRLE strings.Builder
	for ; n > 1; n-- {
		newRLE.Reset()

		count, slen := 1, len(rle)
		for idx := 1; idx < slen; idx++ {
			if rle[idx] != rle[idx-1] {
				newRLE.WriteString(strconv.Itoa(count))
				newRLE.WriteByte(rle[idx-1])

				count = 0
			}

			count++
		}

		newRLE.WriteString(strconv.Itoa(count))
		newRLE.WriteByte(rle[slen-1])
		rle = newRLE.String()
	}

	return rle
}

// // Approach: Recursive
// // Time: O(nxm), m=max(len(RLE))
// // Space: O(n)
// func countAndSay(n int) string {
// 	if n == 1 {
// 		return "1"
// 	}
// 	var s = countAndSay(n - 1)
// 	var rle strings.Builder
// 	var count, slen = 1, len(s)
// 	for idx := 1; idx < slen; idx++ {
// 		if s[idx] != s[idx-1] {
// 			rle.WriteString(strconv.Itoa(count))
// 			rle.WriteByte(s[idx-1])
// 			count = 0
// 		}
// 		count++
// 	}
// 	rle.WriteString(strconv.Itoa(count))
// 	rle.WriteByte(s[slen-1])
// 	return rle.String()
// }
